Crash during parse of rego expression having key of form 'foo.bar'

# Expected Behavior
Loading the following should parse successfully
```
test = { a |
  foo := {"a" : "b"}
  a := { foo.a : "b" }
}
``` 
Code version is OPA 0.14.2 (commit db9855d1, built at 2019-10-04T15:28:47Z)


# Actual Behavior
When opa run is done loading in this expression get
```
panic: corrupt object [recovered]
        panic: corrupt object

goroutine 1 [running]:
github.com/open-policy-agent/opa/ast.(*Compiler).compile.func1()
        /go/src/github.com/open-policy-agent/opa/ast/compile.go:812 +0x73
panic(0xe7bee0, 0x108ca10)
        /usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/open-policy-agent/opa/ast.(*object).Iter(0xc0001b2330, 0xc0001ffa80, 0x20, 0xeeb960)
        /go/src/github.com/open-policy-agent/opa/ast/term.go:1736 +0xf2
github.com/open-policy-agent/opa/ast.(*object).Foreach(0xc0001b2330, 0xc0001af460)
        /go/src/github.com/open-policy-agent/opa/ast/term.go:1759 +0x53
github.com/open-policy-agent/opa/ast.walk(0x109f320, 0xc0001aaa60, 0xf737e0, 0xc0001b2330)
        /go/src/github.com/open-policy-agent/opa/ast/visit.go:110 +0x835
github.com/open-policy-agent/opa/ast.Walk(0x109f320, 0xc0001aaa60, 0xf737e0, 0xc0001b2330)
        /go/src/github.com/open-policy-agent/opa/ast/visit.go:27 +0x15a
github.com/open-policy-agent/opa/ast.walk(0x109f320, 0xc0001aaa60, 0xf53180, 0xc0001a7680)
        /go/src/github.com/open-policy-agent/opa/ast/visit.go:104 +0xe9e
github.com/open-policy-agent/opa/ast.Walk(0x109f320, 0xc0001aaa60, 0xf53180, 0xc0001a7680)
        /go/src/github.com/open-policy-agent/opa/ast/visit.go:27 +0x15a
github.com/open-policy-agent/opa/ast.walk(0x109f320, 0xc0001aaa60, 0xf751e0, 0xc000191770)
        /go/src/github.com/open-policy-agent/opa/ast/visit.go:92 +0x985
github.com/open-policy-agent/opa/ast.Walk(0x109f320, 0xc0001aaa60, 0xf751e0, 0xc000191770)
        /go/src/github.com/open-policy-agent/opa/ast/visit.go:27 +0x15a
github.com/open-policy-agent/opa/ast.walk(0x109f320, 0xc0001aaa60, 0xf54260, 0xc0001af420)
        /go/src/github.com/open-policy-agent/opa/ast/visit.go:80 +0xba6

```
# Steps to Reproduce the Problem

  1. create a refo file with content 
```
test = { a |
  foo := {"a" : "b"}
  a := { foo.a : "b" }
}
```
call this test.rego
  2. opa run test.rego
  3.

# Additional Info

--- a/ast/compile.go
+++ b/ast/compile.go
@@ -3053,9 +3053,16 @@ func rewriteDeclaredVarsInTerm(g *localVarGenerator, stack *localDeclaredVars, t
 	case Object:
 		cpy, _ := v.Map(func(k, v *Term) (*Term, *Term, error) {
 			kcpy := k.Copy()
-			_, errs = rewriteDeclaredVarsInTerm(g, stack, kcpy, errs)
-			_, errs = rewriteDeclaredVarsInTerm(g, stack, v, errs)
+			errs = rewriteDeclaredVarsInTermRecursive(g, stack, kcpy, errs)
+			errs = rewriteDeclaredVarsInTermRecursive(g, stack, v, errs)
 			return kcpy, v, nil
+		})
+		term.Value = cpy
+	case Set:
+		cpy, _ := v.Map(func(elem *Term) (*Term, error) {
+			elemcpy := elem.Copy()
+			errs = rewriteDeclaredVarsInTermRecursive(g, stack, elemcpy, errs)
+			return elemcpy, nil
 		})
 		term.Value = cpy
 	case *ArrayComprehension:
@@ -3072,15 +3079,15 @@ func rewriteDeclaredVarsInTerm(g *localVarGenerator, stack *localDeclaredVars, t
 
 func rewriteDeclaredVarsInTermRecursive(g *localVarGenerator, stack *localDeclaredVars, term *Term, errs Errors) Errors {
 	WalkNodes(term, func(n Node) bool {
+		var stop bool
 		switch n := n.(type) {
 		case *With:
 			_, errs = rewriteDeclaredVarsInTerm(g, stack, n.Value, errs)
-			return true
+			stop = true
 		case *Term:
-			_, errs = rewriteDeclaredVarsInTerm(g, stack, n, errs)
-			return false
+			stop, errs = rewriteDeclaredVarsInTerm(g, stack, n, errs)
 		}
-		return false
+		return stop
 	})
 	return errs
 }

--- a/ast/compile_test.go
+++ b/ast/compile_test.go
@@ -1528,7 +1528,7 @@ func TestCompilerRewriteLocalAssignments(t *testing.T) {
 			`,
 			exp: `
 				package test
-		
+
 				f(x) = __local0__ { x == 1; __local0__ = 2 } else = __local1__ { x == 3; __local1__ = 4 }
 			`,
 			expRewrittenMap: map[Var]Var{
@@ -1635,6 +1635,78 @@ func TestRewriteLocalVarDeclarationErrors(t *testing.T) {
 		if result[i] != expectedErrors[i] {
 			t.Fatalf("Expected:\n\n%v\n\nGot:\n\n%v", strings.Join(expectedErrors, "\n"), strings.Join(result, "\n"))
 		}
+	}
+}
+
+func TestRewriteDecledVarsStage(t *testing.T) {
+
+	// Unlike the following test case, this only executes up to the
+	// RewriteLocalVars stage. This is done so that later stages like
+	// RewriteDynamics are not executed.
+
+	tests := []struct {
+		note   string
+		module string
+		exp    string
+	}{
+		{
+			note: "object ref key",
+			module: `
+				package test
+
+				p {
+					a := {"a": "a"}
+					{a.a: a.a}
+				}
+			`,
+			exp: `
+				package test
+
+				p {
+					__local0__ = {"a": "a"}
+					{__local0__.a: __local0__.a}
+				}
+			`,
+		},
+		{
+			note: "set ref element",
+			module: `
+				package test
+
+				p {
+					a := {"a": "a"}
+					{a.a}
+				}
+			`,
+			exp: `
+				package test
+
+				p {
+					__local0__ = {"a": "a"}
+					{__local0__.a}
+				}
+			`,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.note, func(t *testing.T) {
+
+			c := NewCompiler()
+
+			c.Modules = map[string]*Module{
+				"test.rego": MustParseModule(tc.module),
+			}
+
+			compileStages(c, c.rewriteLocalVars)
+
+			exp := MustParseModule(tc.exp)
+			result := c.Modules["test.rego"]
+
+			if !exp.Equal(result) {
+				t.Fatalf("Expected:\n\n%v\n\nGot:\n\n%v", exp, result)
+			}
+		})
 	}
 }
 

